google/cloud/storage/internal/object_read_streambuf.cc

```diff
@@ -65,100 +65,93 @@ void ObjectReadStreambuf::Close() {
-
-  auto hint = headers_.end();
-  for (auto& kv : read_result->response.headers) {
-    hash_validator_->ProcessHeader(kv.first, kv.second);
-    hint =
-        std::next(headers_.emplace_hint(hint, kv.first, std::move(kv.second)));
-  }
-  if (read_result->response.status_code >= HttpStatusCode::kMinNotSuccess) {
-    return AsStatus(read_result->response);
-  }
-
-  if (!current_ios_buffer_.empty()) {
-    char* data = current_ios_buffer_.data();
-    hash_validator_->Update(data, current_ios_buffer_.size());
-    setg(data, data, data + current_ios_buffer_.size());
-    return traits_type::to_int_type(*data);
-  }
-
-  // This is an actual EOF, there is no more data to download, create an
-  // empty (but valid) region:
-  SetEmptyRegion();
+  status_ = std::move(status);
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+  google::cloud::internal::ThrowStatus(status_);
+#else
   return traits_type::eof();
+#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 }
 
-ObjectReadStreambuf::int_type ObjectReadStreambuf::underflow() {
-  auto next_char = Peek();
-  if (!next_char) return ReportError(next_char.status());
-  if (*next_char != traits_type::eof()) return *next_char;
-
-  auto msg = FinishValidator(__func__);
-  if (!hash_validator_result_.is_mismatch) return traits_type::eof();
+void ObjectReadStreambuf::ThrowHashMismatchDelegate(const char* function_name) {
+  std::string msg;
+  msg += function_name;
+  msg += "(): mismatched hashes in download";
+  msg += ", computed=";
+  msg += computed_hash();
+  msg += ", received=";
+  msg += received_hash();
   if (status_.ok()) {
     // If there is an existing error, we should report that instead because
-    // it is more specific. For example, every permanent network error will
-    // produce invalid checksums, but that is not the interesting
-    // information.
+    // it is more specific, for example, every permanent network error will
+    // produce invalid checksums, but that is not the interesting information.
     status_ = Status(StatusCode::kDataLoss, msg);
   }
 #if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+  // The only way to report errors from a std::basic_streambuf<> (which this
+  // class derives from) is to throw exceptions:
+  //   https://stackoverflow.com/questions/50716688/how-to-set-the-badbit-of-a-stream-by-a-customized-streambuf
+  // but we need to be able to report errors when the application has
+  // disabled exceptions via `-fno-exceptions` or a similar option. In that
+  // case we set `status_`, and report the error as an 0-byte read. This is
+  // obviously not ideal, but it is the best we can do when the application
+  // disables the standard mechanism to signal errors.
   throw HashMismatchError(msg, received_hash(), computed_hash());
-#else
-  return traits_type::eof();
 #endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 }
 
+bool ObjectReadStreambuf::ValidateHashes(char const* function_name) {
+  auto validator = std::move(hash_validator_);
```

Comment:
It's interesting that calling `ValidateHashes()` "consumes" the `hash_validator_` member. I assume there's some guarantee about the way that this code (or streams in general) work that ensures that this function cannot be called twice or that any other functions that use `hash_validator_` can be called after this function. If there's a link to some docs that explain this it may be worth including here. Otherwise, even a small comment like "// `hash_validator_` cannot be used after this function because ...".

Reply:
Done. FWIW, we were already consuming `*hash_validator_`.  The call to `std::move(*hash_validator_).Finish()` effectively made the validator unusable, it was just harder to see.

Reply:
I see that now. The code is indeed clearer now. And thanks for the comment.